
### Classification using a Simple Neural Network on MNIST Dataset.


In this example, we will use the TensorFlow library to create a simple neural network-based classification model for the MNIST dataset. The MNIST dataset consists of handwritten digits (0-9) and is a popular benchmark dataset for image classification tasks.

In [2]:
# Install the required libraries:
!pip install tensorflow numpy

In [3]:
#Importing the necessary libraries:

import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical


In [4]:
# Loading and preprocessing the MNIST dataset:

# Load the MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalizing the pixel values to be between 0 and 1
train_images = train_images / 255.0
test_images = test_images / 255.0

# Converting the labels to one-hot encoded vectors
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


11490434/11490434 [==============================] - 3s 0us/step


The dataset is split into a training set (60,000 samples) and a test set (10,000 samples). We normalize the pixel values to be between 0 and 1 and convert the labels to one-hot encoded vectors for classification.

In [5]:
# Definning the neural network model:

# Create a sequential model
model = Sequential()

# Adding a Flatten layer to convert 2D images to 1D arrays
model.add(Flatten(input_shape=(28, 28)))

# Adding a Dense hidden layer with 128 units and a ReLU activation function
model.add(Dense(128, activation='relu'))

# Adding a Dense output layer with 10 units (for 10 classes) and a softmax activation function
model.add(Dense(10, activation='softmax'))


We create a sequential model and add three layers: a Flatten layer to convert the 2D images (28x28 pixels) into 1D arrays, a Dense hidden layer with 128 units and a ReLU activation function, and a Dense output layer with 10 units (for the 10 classes) and a softmax activation function.

In [6]:
# Compiling the model:

model.compile(optimizer='adam', loss='categorical_crossentropy',
              metrics=['accuracy'])


We compile the model using the Adam optimizer, categorical cross-entropy loss, and accuracy metric.

In [7]:
# Training the model:

model.fit(train_images, train_labels, epochs=10, batch_size=32,
          validation_split=0.1)


Epoch 1/10
1688/1688 [==============================] - 8s 4ms/step - loss: 0.2792 - accuracy: 0.9207 - val_loss: 0.1242 - val_accuracy: 0.9662
Epoch 2/10
1688/1688 [==============================] - 8s 5ms/step - loss: 0.1241 - accuracy: 0.9628 - val_loss: 0.0946 - val_accuracy: 0.9723
Epoch 3/10
1688/1688 [==============================] - 7s 4ms/step - loss: 0.0861 - accuracy: 0.9741 - val_loss: 0.0950 - val_accuracy: 0.9717
Epoch 4/10
1688/1688 [==============================] - 7s 4ms/step - loss: 0.0630 - accuracy: 0.9811 - val_loss: 0.0785 - val_accuracy: 0.9768
Epoch 5/10
1688/1688 [==============================] - 7s 4ms/step - loss: 0.0495 - accuracy: 0.9853 - val_loss: 0.0764 - val_accuracy: 0.9785
Epoch 6/10
1688/1688 [==============================] - 7s 4ms/step - loss: 0.0375 - accuracy: 0.9883 - val_loss: 0.0787 - val_accuracy: 0.9800
Epoch 7/10
1688/1688 [==============================] - 7s 4ms/step - loss: 0.0304 - accuracy: 0.9907 - val_loss: 0.0860 - val_accuracy:

We train the model using the training data for 10 epochs with a batch size of 32 and a validation split of 10% to monitor the model's performance on unseen data during training.

In [8]:
# Evaluating the model:

test_loss, test_accuracy = model.evaluate(test_images, test_labels)
print(f'Test accuracy: {test_accuracy:.4f}')


313/313 [==============================] - 1s 2ms/step - loss: 0.0751 - accuracy: 0.9791
Test accuracy: 0.9791


We evaluate the model on the test dataset and print the test accuracy.

That's it! You now have a simple neural network-based classification model for the MNIST dataset. The model achieves a test accuracy of around 98%. We can further improve the model by adding more layers, using dropout or batch normalization, or adjusting the hyperparameters.